core/pkg/store/store.go

```diff
@@ -4,66 +4,117 @@ import (
 	"context"
 	"encoding/json"
 	"fmt"
-	"maps"
-	"reflect"
+	"slices"
 	"sync"
 
 	"github.com/hashicorp/go-memdb"
 	"github.com/open-feature/flagd/core/pkg/logger"
 	"github.com/open-feature/flagd/core/pkg/model"
+	"github.com/open-feature/flagd/core/pkg/utils"
 )
 
-type del = struct{}
+var noValidatedSources = []string{}
 
-var deleteMarker *del
+type SelectorContextKey struct{}
+
+// do we really need this?
+type Payload struct {
+	Flags map[string]model.Flag
+}
 
 type IStore interface {
-	GetAll(ctx context.Context) (map[string]model.Flag, model.Metadata, error)
-	Get(ctx context.Context, key string) (model.Flag, model.Metadata, bool)
-	SelectorForFlag(ctx context.Context, flag model.Flag) string
+	GetAll(ctx context.Context, selector *Selector, watcher chan Payload) (map[string]model.Flag, model.Metadata, error)
+	Get(ctx context.Context, key string, selector *Selector) (model.Flag, model.Metadata, bool)
 }
 
 type Store struct {
-	mx                sync.RWMutex
-	db                *memdb.MemDB
-	logger            *logger.Logger
-	FlagSources       []string
-	SourceDetails     map[string]SourceDetails  `json:"sourceMetadata,omitempty"`
-	MetadataPerSource map[string]model.Metadata `json:"metadata,omitempty"`
+	mx      sync.RWMutex
+	db      *memdb.MemDB
+	logger  *logger.Logger
+	sources []string
+	// deprecated: has no effect and will be removed soon.
+	FlagSources []string
 }
 
 type SourceDetails struct {
 	Source   string
 	Selector string
 }
 
-func (f *Store) hasPriority(stored string, new string) bool {
-	if stored == new {
-		return true
-	}
-	for i := len(f.FlagSources) - 1; i >= 0; i-- {
-		switch f.FlagSources[i] {
-		case stored:
-			return false
-		case new:
-			return true
-		}
-	}
-	return true
-}
-
-func NewStore(logger *logger.Logger) (*Store, error) {
+func NewStore(logger *logger.Logger, sources []string) (*Store, error) {
 
+	// a unique index must exist for each set of constraints - for example, to look up by key and source, we need a compound index on key+source, etc
+	// we maybe want to generate these dynamically in the future to support more robust querying, but for now we will hardcode the ones we need
```

Comment:
As the comment here mentions, we need a distinct index for every possible combination of constraint arguments. If we move beyond `selector` and `source`, we will probably need to generate these dynamically, but these are all that's needed for now.